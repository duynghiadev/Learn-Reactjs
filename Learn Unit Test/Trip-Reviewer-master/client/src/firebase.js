import * as firebase from 'firebase';

const config = {
  apiKey: 'AIzaSyBLEX7iZntFl-xVHhe9CRRRQiAxHnW1dWg',
  authDomain: 'trip-reviewer-2018.firebaseapp.com',
  databaseURL: 'https://trip-reviewer-2018.firebaseio.com',
  projectId: 'trip-reviewer-2018',
  storageBucket: 'trip-reviewer-2018.appspot.com',
  messagingSenderId: '721559559011',
};

firebase.initializeApp(config);

export { firebase };
