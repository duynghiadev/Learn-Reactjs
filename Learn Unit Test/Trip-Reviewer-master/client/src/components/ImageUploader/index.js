import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CustomUploadButton from 'react-firebase-file-uploader/lib/CustomUploadButton';
import { firebase } from '../../firebase';

class ImageUploader extends Component {
  componentWillMount() {
    firebase.auth().signInAnonymously().catch(this.handleAuthError);
  }

  componentWillUnmount() {
    firebase.auth().signOut();
  }

  handleAuthError = error => {
    console.log(error); // eslint-disable-line no-console
  };

  handleUploadStart = () => {
    this.props.onUploadStarts();
  };

  handleUploadError = error => {
    console.log(error); // eslint-disable-line no-console
    this.props.onUploadError(error);
  };

  handleProgress = progress => {
    this.props.onUploadProgress(progress);
  };

  handleUploadSuccess = filename => {
    this.props.onUploadSuccess(filename);
  };

  render() {
    return (
      <CustomUploadButton
        hidden
        accept="image/*"
        name="avatar"
        filename={this.props.filename}
        metadata={{ cacheControl: 'max-age=86400' }}
        storageRef={firebase.storage().ref('avatars')}
        onUploadStart={this.handleUploadStart}
        onUploadError={this.handleUploadError}
        onUploadSuccess={this.handleUploadSuccess}
        onProgress={this.handleProgress}
        className={this.props.className}
      >
        {this.props.children}
      </CustomUploadButton>
    );
  }
}

ImageUploader.propTypes = {
  children: PropTypes.node.isRequired,
  filename: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
  onUploadStarts: PropTypes.func.isRequired,
  onUploadError: PropTypes.func.isRequired,
  onUploadProgress: PropTypes.func.isRequired,
  onUploadSuccess: PropTypes.func.isRequired,
};

export default ImageUploader;
