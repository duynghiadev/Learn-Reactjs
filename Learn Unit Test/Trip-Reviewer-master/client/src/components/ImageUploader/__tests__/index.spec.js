import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';

import ImageUploader from '..';
import { firebase } from '../../../firebase';

jest.mock('react-firebase-file-uploader/lib/CustomUploadButton', () => 'CustomUploadButton');
jest.mock('./../../../firebase', () => ({
  firebase: {
    auth: jest.fn(() => ({
      signInAnonymously: jest.fn(() => ({
        catch: jest.fn(),
      })),
      signOut: jest.fn(),
    })),
    storage: jest.fn(() => ({
      ref: jest.fn(),
    })),
  },
}));

const mockComponent = props => <ImageUploader {...props}>Test</ImageUploader>;

describe('<ImageUploader />', () => {
  it('should render component', () => {
    const tree = create(mockComponent({})).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should change state when handleUploadStart is called', () => {
    const onUploadStarts = jest.fn();
    const props = { onUploadStarts };
    const instance = shallow(mockComponent(props)).instance();

    instance.handleUploadStart();

    expect(onUploadStarts).toHaveBeenCalled();
  });

  it('should change state when handleUploadError is called', () => {
    const onUploadError = jest.fn();
    const props = { onUploadError };
    const instance = shallow(mockComponent(props)).instance();

    instance.handleUploadError();

    expect(onUploadError).toHaveBeenCalled();
  });

  it('should change state when handleProgress is called', () => {
    const onUploadProgress = jest.fn();
    const props = { onUploadProgress };
    const instance = shallow(mockComponent(props)).instance();

    instance.handleProgress(20);

    expect(onUploadProgress).toHaveBeenCalledWith(20);
  });

  it('should change state when handleAuthError is called', () => {
    const instance = shallow(mockComponent({})).instance();

    instance.handleAuthError('error');
    // TODO: there is nothing except the console.log there at the moment...
  });

  it('should change state when handleUploadSuccess and call onUploadSuccess is called', () => {
    const onUploadSuccess = jest.fn();
    const props = { onUploadSuccess };
    const instance = shallow(mockComponent(props)).instance();

    instance.handleUploadSuccess('filename');

    expect(onUploadSuccess).toHaveBeenCalledWith('filename');
  });

  it('should call firebase.signOut() when componentWillUnmount is called', () => {
    const instance = shallow(mockComponent()).instance();

    instance.componentWillUnmount();

    expect(firebase.auth).toHaveBeenCalled();
  });
});
