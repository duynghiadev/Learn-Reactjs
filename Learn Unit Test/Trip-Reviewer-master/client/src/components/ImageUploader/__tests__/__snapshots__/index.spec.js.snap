// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageUploader /> should render component 1`] = `
<CustomUploadButton
  accept="image/*"
  className={undefined}
  filename={undefined}
  hidden={true}
  metadata={
    Object {
      "cacheControl": "max-age=86400",
    }
  }
  name="avatar"
  onProgress={[Function]}
  onUploadError={[Function]}
  onUploadStart={[Function]}
  onUploadSuccess={[Function]}
  storageRef={undefined}
>
  Test
</CustomUploadButton>
`;
